azurelinuxagent/common/protocol/extensions_goal_state_from_extensions_config.py

```diff
@@ -0,0 +1,559 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+import json
+
+from collections import defaultdict
+
+import azurelinuxagent.common.logger as logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.exception import ExtensionsConfigError
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.protocol.extensions_goal_state import ExtensionsGoalState
+from azurelinuxagent.common.protocol.restapi import Extension, ExtHandler, ExtHandlerVersionUri, VMAgentManifest, \
+    VMAgentManifestUri, ExtensionState, ExtHandlerList, VMAgentManifestList, InVMGoalStateMetaData
+from azurelinuxagent.common.utils.textutil import parse_doc, parse_json, findall, find, findtext, getattrib, gettext, format_exception, \
+    is_str_none_or_whitespace, is_str_empty
+
+
+class ExtensionsGoalStateFromExtensionsConfig(ExtensionsGoalState):
```

Comment:
Moved from extensions_goal_state.py, with a few changes to expose some methods as properties instead.

Reply:
Could this class (and the VMSettings one) just be private, inside e.g. `common/protocol/extensions_goal_state_factory.py`? Seems like exporting the `@property`s via the base class should be the only thing that needs to be public.

Reply:
They used to be private, but there are a couple of cases where the code needs to be aware of the concrete class it is dealing with (to, for example, send telemetry about the HostGAPlugin) so I made them public.